move pkg to cmds

rajatjindal opened the pull request:
an ambitious try to move commands to a different package then 'main' to be able to use them as library.

zetaab commented:
/ok-to-test

chrisz100 commented:
I'm wondering what your goal with publishing the commands as a library is? That's quite a huge change...

rajatjindal commented:
Hi @chrisz100 

do you think this change can be merged. If not we can close the PR and I will look for other ways to do what I am trying to do.

If you think it can be merged, I will rebase it against master.

Many thanks for considering this PR.

Rajat Jindal

chrisz100 commented:
@rajatjindal makes sense - by the way no need to be so formal. It's all just people eager to move this project forward here :-) 

It's quite a huge PR, do you have some information on how you tested everything kops does still works?

chrisz100 commented on README.md:
```diff
@@ -1,4 +1,4 @@
-<img src="/docs/img/logo.jpg" width="500px" alt="kops logo">
+<img src="/docs/img/logo.jpg" width="500px" alt="kops logo"> 
```
Don't really see this as a change that should be in this PR

rajatjindal replied:
i think it was to trigger ci again as it flaked on a test.

chrisz100 commented on cmd/kops/BUILD.bazel:
```diff
@@ -1,131 +1,11 @@
-load("@io_bazel_rules_go//go:def.bzl", "go_binary", "go_library", "go_test")
+load("@io_bazel_rules_go//go:def.bzl", "go_binary", "go_library")
```
I'm not an expert in bazel. @justinsb is this good to remove?

rajatjindal replied:
i think this is generated. i didn't explicitly removed it.

drekle commented:
This kind of steps away from a traditional folder structure that you see in most go projects.  When I am looking at a project I expect most libraries to be under `/pkg`.  While there is no universally accepted standard for this, it seems common in most of the repos.

Consider:
https://github.com/golang-standards/project-layout

Also maybe refer to how its done here, which I find a little cleaner: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubectl/kubectl.go#L29

I don't the PR is that ambitious really.  Just in the wrong place.

drekle commented on cmd/kops/cmds/create_cluster.go:
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 */
 
-package main
+package cmds
```
Consider the following layout instead.

https://github.com/kubernetes/kubernetes/blob/master/cmd/kubectl/kubectl.go#L29

chrisz100 replied:
Generally agree but the kops project is so spread I doubt starting with this would help to clean this up. @mikesplain @justinsb what do you say?

rajatjindal replied:
requesting feedback from @mikesplain @justinsb

chrisz100 replied:
As this line refers to the package name and does not import anything this looks perfectly fine to me. 

All projects do it as proposed here, see: 
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubectl/kubectl.go#L17

zetaab commented:
from my perspective this PR is really interesting and I would like to see progress in this. How we could get this forward? I know its big change, but this allows use kops as library. Now the situation is that if we want use kops as library we need copy paste quite much stuff.

zetaab commented:
@justinsb @mikesplain could you please give your opinion here? At least for me this PR looks very interesting because I would like to use kops as library from another code.

chrisz100 commented:
Can you please rebase and fix the tests @rajatjindal ? Also, is this WIP still? When do you plan to have this ready?

rajatjindal commented:
I wanted to make sure this PR is acceptable to other core members of the project. I will rebase shortly and look into the test failure as well. thanks.

zetaab commented:
@rajatjindal at least I want this PR. I am planning use kops as part of webservice

rajatjindal commented:
Hi @zetaab @chrisz100 

I've rebased the PR and fixed the tests. can you please sponsor this PR with other core members of the project and help it move forward.

Many thanks
Rajat Jindal

justinsb commented:
This seems fine in general, though it's going to be hard to get it to a point where there are no conflicts, and we do have to prioritize bug fixes and the like.

I would like to see this though - I'm really pushing to get 1.12.0-alpha.1 done, and then hopefully things will settle down a little bit and then it won't be a constant rebase battle.

rajatjindal commented:
Hi @justinsb do you think this PR stand a chance to get merged now if I rebase it against master now?
